SmartDeviceLink/private/SDLIAPSession.m

```diff
@@ -11,101 +11,285 @@
+    return [self.eaSession.outputStream write:buffer maxLength:len];
+}
 
-#pragma mark - Private Stream Lifecycle Helpers
+#pragma mark - Private Stream Helpers
 
-- (BOOL)createSession {
-    SDLLogD(@"Opening EASession with accessory: %@", self.accessory.name);
-    self.eaSession = [[EASession alloc] initWithAccessory:self.accessory forProtocol:self.protocolString];
-    return (self.eaSession != nil);
+- (void) startSession {
+    dispatch_async(self.iapSessionQueue, ^{
+        self.eaSession = [[EASession alloc] initWithAccessory:self.accessory forProtocol:self.protocolString];
+        SDLLogD(@"Created EASession with %@ Protocol and EASession is %@", self.protocolString, self.eaSession);
+        if (self.eaSession != nil) {
+            [self openStreams];
+            [self startStreamRunLoop];
+        } else {
+            SDLLogD(@"Failed to create EASession with Protocol : %@", self.protocolString);
+        }
+    });
 }
 
-- (void)startStream:(NSStream *)stream {
-    stream.delegate = self;
-    [stream scheduleInRunLoop:[NSRunLoop currentRunLoop] forMode:NSDefaultRunLoopMode];
-    [stream open];
+- (void) openStreams {
+    if (self.eaSession != nil) {
+        [[self.eaSession inputStream] setDelegate: self];
+        [[self.eaSession inputStream] scheduleInRunLoop:[NSRunLoop currentRunLoop] forMode:NSDefaultRunLoopMode];
+        [[self.eaSession inputStream] open];
+        [[self.eaSession outputStream] setDelegate: self];
+        [[self.eaSession outputStream] scheduleInRunLoop:[NSRunLoop currentRunLoop] forMode:NSDefaultRunLoopMode];
+        [[self.eaSession outputStream] open];
+    } else {
+        SDLLogD(@"EASession is nil when calling openStreams()");
+    }
 }
 
-- (void)stopStream:(NSStream *)stream {
-    // Verify stream is in a state that can be closed. Closing a stream that has not been opened has very bad effects.
-    NSUInteger status1 = stream.streamStatus;
-    if (status1 != NSStreamStatusNotOpen &&
-        status1 != NSStreamStatusClosed) {
-        [stream close];
-    } else if (status1 == NSStreamStatusNotOpen) {
-        // It's implicitly removed from the stream when it's closed, but not if it was never opened.
-        // When the USB cable is disconnected, the app will will call this method after the `NSStreamEventEndEncountered` event. The stream will already be in the closed state but it still needs to be removed from the run loop.
-        [stream removeFromRunLoop:[NSRunLoop currentRunLoop] forMode:NSDefaultRunLoopMode];
+- (void)close:(NSStream *)stream {
+    if (stream.streamStatus == NSStreamStatusClosed) {
+        if ([self isInputStream:stream]) {
+            SDLLogD(@"EASession inputstream already closed for EASession %@", self.eaSession);
+        }
+        if ([self isOutputStream:stream]) {
+            SDLLogD(@"EASession outputstream already closed for EASession %@", self.eaSession);
+        }
+        return;
     }
-
+    
+    [stream close];
```

Comment:
While testing, I have experienced random crashes at `[stream close]`. I suspect the issue is that `[stream close]` is being called on the main thread due to the accessory disconnected notification being sent on the main thread (in the `SDLIAPTransport` transport class). We have experienced crash in the [past](https://github.com/smartdevicelink/sdl_ios/pull/1592) and have found that we must open and close the I/O streams on the same thread. 

Since you switched from using a thread to dispatching to the `iapSessionQueue`, this means we have no guarantee that the I/O streams will be opened and closed on the same thread even if we change the code to dispatch closing the I/O threads to the `iapSessionQueue` (due to the fact that behind the scenes the blocks can be dispatched to different threads).

I was curious what the motivation was behind the switching from using a thread to to using the `iapSessionQueue`? Obviously, your changes are simpler and easier to read so I hope there is a solution that lets us keep the  `iapSessionQueue`.

Reply:
iOS recommends that new designs not use NSThread and instead to use newer techologies such as GCD.  Using a DispatchQueue moves the responsility of creating and managing the thread to iOS. As you correctly mention, dispatch queue blocks can be dispatched to different threads. However self.eaSession always runs on the same thread, the thread that GCD used when we create self.eaSession here in SDLIAPSession

 (void) startSession {
dispatch_async(self.iapSessionQueue, ^{
        self.eaSession = [[EASession alloc] initWithAccessory:self.accessory forProtocol:self.protocolString];

Notice there are no other calls to dispatch_async(self.iapSessionQueue) anywhere in SDLIAPSession.  All further access is via self.eaSession, thus [stream close] is always on the same thread as was used for example in [stream open]

Reply:
@kmicha19-ford I'm not quite following. Are you saying that `self.eaSession` is internally thread-safe and any calls to it will internally dispatch to the correct thread, and therefore it doesn't matter from where the `NSStream close` call comes from? Since those seem to be separate objects and classes, could you clarify the relation? I know an `EASession` uses the stream, but you seem to be closing the stream without accessing `self.eaSession` at this code line. As @NicoleYarroch noted, there seem to be crashes with this PR related to closing the stream, so it can't be entirely safe here.

Reply:
Sorry for delay with answer.  Work Mac went down yesterday. IT scheduled an appointment for Monday to take it in to run diagnostics. Meanwhile I am on a backup computer. 

EASession is not thread safe. As to other questions I need to get things setup on this backup Mac and so I can run this code and look at some things before answering these questions. Hopefully by Monday if not end of day today.

Reply:
@NicoleYarroch - Are you seeing random crashes at [stream close] with this PR? (Or something with existing code base)?

Reply:
@joeljfischer Which separate objects and classes are you referring to in your comment?

Reply:
I was testing with the Swift example app using your branch. The crash occurred when I pulled the USB cord. I was testing with an iPhone 11 (iOS v.132.1). The crash only happened twice during ~30 disconnect tests. 

It was crashing at `[stream close]` both times. When I looked at the stack trace, I could see that the main thread was executing `[stream close]`. I traced it back to getting the accessory disconnect notification. 

I did notice while testing that sometimes `[stream close]` was triggered by the stream getting an `NSStreamEventEndEncountered` event and other times by the accessory disconnect notification. Sometimes both threads were trying to close the stream. It could be that the crash was related to both threads trying to close the stream at the same time? However I suspect that it is more likely that the the stream was being closed by a thread that does not own the stream (i.e. the main thread).

Reply:
I have a fix that ensures sessions are always closed on the sessions thread.  However there is an issue pushing the commit from my backup Mac. I am working on resolving this now.

Reply:
@NicoleYarroch - I pushed a fix for close session.   Review appreciated.